crates/sawtooth-tp/src/opa.rs

```diff
@@ -0,0 +1,141 @@
+use async_sawtooth_sdk::protobuf::Message;
+use chronicle_protocol::settings::sawtooth_settings_address;
+use common::opa::{CliPolicyLoader, ExecutorContext};
+use sawtooth_sdk::{
+    messages::setting::Setting,
+    processor::handler::{ApplyError, TransactionContext},
+};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use tracing::{debug, info, warn};
+
+#[derive(Debug)]
+pub struct TpOpa {
+    pub embedded: ExecutorContext,
+    pub on_chain: Arc<Mutex<HashMap<String, (String, ExecutorContext)>>>,
+}
+
+impl TpOpa {
+    pub fn new(policy: &str, entrypoint: &str) -> Result<Self, ApplyError> {
+        Ok(Self {
+            on_chain: Arc::new(HashMap::new().into()),
+            embedded: {
+                ExecutorContext::from_loader(
+                    &CliPolicyLoader::from_embedded_policy(policy, entrypoint)
+                        .map_err(|e| ApplyError::InternalError(e.to_string()))?,
+                )
+                .map_err(|e| ApplyError::InternalError(e.to_string()))?
+            },
+        })
+    }
+
+    pub fn executor_context(
+        &self,
+        ctx: &mut dyn TransactionContext,
+    ) -> Result<ExecutorContext, ApplyError> {
+        let policy_name_settings_entry =
+            ctx.get_state_entry(&sawtooth_settings_address("chronicle.opa.policy_name"))?;
+
+        let entrypoint_settings_entry =
+            ctx.get_state_entry(&sawtooth_settings_address("chronicle.opa.entrypoint"))?;
+
+        match (policy_name_settings_entry, entrypoint_settings_entry) {
+            (None, None) => {
+                warn!("Insecure operating mode - no on chain policy name or entrypoint settings found");
+                Ok(self.embedded.clone())
+            }
+            (Some(policy_name_settings_entry), Some(entrypoint_settings_entry)) => {
+                info!("Chronicle operating in secure mode - on chain policy name and entrypoint settings found");
+                let policy_name_settings_entry: Setting =
+                    Message::parse_from_bytes(&policy_name_settings_entry).map_err(|_e| {
```

Comment:
With these various internal errors, ought we helpfully log the entry and (where there is one) the underlying error?